### PR TITLE
Update Dockerfile with best practices

### PR DESCRIPTION
**Pull request by nate-droid**
This PR will:

> Add an unprivileged user (remove using root)
> Reduce the size of the image by using a distroless scratch
>   > as a side effect, this reduces the attack surface
>   > added multistage build

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,31 +1,47 @@
...
+    --disabled-password \
+    --gecos "" \
+    --home "/nonexistent" \
+    --shell "/sbin/nologin" \
+    --no-create-home \
+    --uid "${UID}" \
+    "${USER}"
 
-WORKDIR /src
+WORKDIR $GOPATH/src/main/bubbly/
```
Comment (olliefr) [inline Dockerfile:17]: Nit: the notation makes me think the distro uses `GOPATH` in its original way [1]. If so, should this not be a fully-qualified name?

As in `$GOPATH/src/github.com/valocode/bubbly`?

This is a nit and I'm just wondering, not necessarily requesting change.

[1]: [How to Write Go Code (with GOPATH) | golang.org](https://golang.org/doc/gopath_code)

File: Dockerfile [hunk 2]
```diff
@@ -1,31 +1,47 @@
...
+RUN go mod verify
 
 COPY . .
 
 RUN go get -u github.com/swaggo/swag/cmd/swag
 RUN swag init
 
-# compile the go binary
-RUN go build -o /bubbly
+RUN GOOS=linux GOARCH=amd64 go build -ldflags="-w -s" -o /go/bin/bubbly
```
Comment (olliefr) [inline Dockerfile:30]: I _suspect_ that `/go` is `$GOPATH` here, judging by the placement of source (my comment above).

File: Dockerfile [hunk 3]
```diff
@@ -1,31 +1,47 @@
...
+ENV USER=bubbly
+ENV UID=10001
+RUN adduser \
+    --disabled-password \
+    --gecos "" \
+    --home "/nonexistent" \
+    --shell "/sbin/nologin" \
+    --no-create-home \
+    --uid "${UID}" \
+    "${USER}"
```
Comment (olliefr) [inline Dockerfile:15]: This does not look right - are you creating a new user in the _builder_ image? I suggest that the user is created in the deployment image, instead.

File: Dockerfile [hunk 4]
```diff
@@ -1,31 +1,47 @@
...
+RUN GOOS=linux GOARCH=amd64 go build -ldflags="-w -s" -o /go/bin/bubbly
 
-# target for running in cluster
-FROM debian:buster-slim
-COPY --from=builder /bubbly /bubbly
+# step 2
+FROM scratch
+# Import the user and group files from the builder.
+COPY --from=builder /etc/passwd /etc/passwd
+COPY --from=builder /etc/group /etc/group
```
Comment (olliefr) [inline Dockerfile:36]: This does not look right 😮 There will be all sorts of service account information there, which may change from release to release. There got to be a better way to do this.

Now I see why you cannot run `useradd` here, it's a `scratch` image, but perhaps the easiest solution is just to append a line to each `/etc/passwd` and `/etc/group`, like Istio does [here](https://github.com/istio/istio/blob/master/docker/Dockerfile.distroless)?

I'm not sure how to best approach this. Moar research is needed 🙃

**Review by olliefr — CHANGES_REQUESTED** [review]
Nice try 😅 but there is a couple of items that I do not feel comfortable with. Having said that, I don't know what is the right thing to do either, but I've left suggestions where I could. Happy to discuss more, ping me on Slack, if the need be.

**Comment by olliefr** [comment]
@nate-droid I just had an idea... what if we deploy it with `distroless` as that Sysdig article suggests? 

✔️ Their smallest image, [`static-debian10`](https://github.com/GoogleContainerTools/distroless/blob/master/base/README.md) has the CA certs baked in. 
✔️ We can follow Istio's approach and echo two lines into `/etc/passwd` and `/etc/group`.
✔️ Then you only need to copy the `bubbly` binary... and we are off to the races! 🙃

File: Dockerfile [hunk 5]
```diff
@@ -1,31 +1,47 @@
+# Stage 1 build
 FROM golang:1.16.1-buster AS builder
```
[the remark this answers is not in the document]
Comment (jlarfors) [reply Dockerfile:2]: I agree with Oliver

File: Dockerfile [hunk 6]
```diff
@@ -1,31 +1,47 @@
...
+RUN apt-get update && apt-get install ca-certificates -y
```
[the remark this answers is not in the document]
Comment (jlarfors) [reply Dockerfile:3]: This is the snippet I usually use for things like this (if interested):

```Dockerfile
RUN apt-get update \
    && apt-get install -y -qq --no-install-recommends ca-certificates \
    && rm -rf /var/lib/apt/lists/*
```

It makes it less verbose in the log and does the cleanup of apt.

File: Dockerfile (hunk 1 above)
Comment (nate-droid) [reply Dockerfile:17]: Good point! Always good to be more precise

File: Dockerfile (hunk 4 above)
Comment (nate-droid) [reply Dockerfile:36]: Looks like distroless contains a user called "nonroot", this completely removes the need to copy anything 😄

File: Dockerfile [hunk 7]
```diff
@@ -1,31 +1,47 @@
...
+ENV CGO_ENABLED=0
```
[the remark this answers is not in the document]
Comment (nate-droid) [reply Dockerfile:4]: This was from when I was playing around with alpine. Not needed anymore, good catch!

File: Dockerfile (hunk 6 above)
[the remark this answers is not in the document]
Comment (nate-droid) [reply Dockerfile:3]: distroless has ca certs built in, so I will just remove this line altogether 😄

**Review by olliefr — APPROVED** [review]
SWEET 🙃

**Review by jlarfors — APPROVED** [review]
LGTM

**Pull request merged**